README.md
# wordle-solver-haskell

[![Haskell CI](https://github.com/siadat/wordle-solver-haskell/actions/workflows/haskell.yml/badge.svg)](https://github.com/siadat/wordle-solver-haskell/actions/workflows/haskell.yml)

A Wordle solver written in Haskell. This tool suggests words (eg `boxer`) and then you provide the colors for that word (eg `02022`).
No heuristics are used. The tool only suggests the first word that complies with previous seen colors and guesses.

I am learning Haskell and I would highly appreciate your feedback!
The [main logic](https://github.com/siadat/wordle-solver-haskell/blob/main/src/Lib.hs#L131-L142) and its [tests](https://github.com/siadat/wordle-solver-haskell/blob/main/test/Spec.hs#L30-L34) might be interesting.

## Usage

Open a Wordle game (eg [Wordle](https://www.powerlanguage.co.uk/wordle/) or [Absurdle](https://qntm.org/files/wordle/index.html)), then run:
```
make run
```

```
Guide:
0     = black
1     = yellow
2     = green
Example:
22001 = 2 greens, 2 blacks, 1 yellow

Enter colors for 'showy': 00000
Enter colors for 'angle': 00001
Enter colors for 'tepid': 01010
Enter colors for 'fixer': 11021
Enter colors for 'brief': 22222
Found!
Bye
```

## Run the tests

```
make test
```

## Acknowledgements
- Thanks @Zij-IT for improving the code via https://github.com/siadat/wordle-solver-haskell/pull/1


ChangeLog.md
# Changelog for wordle-solver

## Unreleased changes


test/Spec.hs
import Test.Hspec
import Lib

import qualified Data.Map as DataMap

main :: IO ()
main = hspec $ do
    describe "judger" $ do
        it "should judge word" $ do
            judgeWord (Master "abcde") (Guess "abcde") `shouldBe` (Colors "22222")
            judgeWord (Master "abcde") (Guess "abxyc") `shouldBe` (Colors "22001")
            judgeWord (Master "a") (Guess "abcde") `shouldBe` (Colors "2")
            judgeWord (Master "abcde") (Guess "a") `shouldBe` (Colors "2")
        it "should guess a word (brute-force)" $ do
            simpleGuesser ["abcde", "abcxy"] DataMap.empty `shouldBe` Just (Guess "abcde")
            simpleGuesser ["abcde", "abcxy"] (DataMap.fromList [(Guess "abcde", Colors "00000")]) `shouldBe` Just (Guess "abcxy")
            simpleGuesser ["abcde", "abcxy"] (DataMap.fromList [(Guess "abcxy", Colors "00000")]) `shouldBe` Just (Guess "abcde")
            simpleGuesser ["abcde", "abcxy"] (DataMap.fromList [(Guess "abcde", Colors "00000"), (Guess "abcxy", Colors "00000")]) `shouldBe` Nothing
        it "...stuff..." $ do
            maskFunc (Guess "abc") (Colors "120") (\x -> x == '2') `shouldBe` "_b_"
            maskFunc (Guess "abc") (Colors "120") (\x -> x /= '2') `shouldBe` "a_c"
            checkMask "a__" "abc" (==) `shouldBe` True
            checkMask "b__" "abc" (==) `shouldBe` False
            checkMask "a__" "abc" (/=) `shouldBe` False
            checkMask "b__" "abc" (/=) `shouldBe` True
            filterGuessByColor (=='1') (Guess "abc") (Colors "121") `shouldBe` "ac"
            filterGuessByColor (=='2') (Guess "abc") (Colors "002") `shouldBe` "c"
            countCharFreq "hello" `shouldBe` DataMap.fromList [('h', 1), ('e', 1), ('l', 2), ('o', 1)]

            matchWord (Guess "hello") (Colors "22222") "hello" `shouldBe` True
            matchWord (Guess "abc") (Colors "210") "axb" `shouldBe` True --
            matchWord (Guess "abc") (Colors "210") "abc" `shouldBe` False
            matchWord (Guess "bc") (Colors "11") "ab" `shouldBe` False
            matchWord (Guess "bc") (Colors "11") "cb" `shouldBe` True
        it "should guess a word (better)" $ do

            let twoLetterWithoutRepeat = ["ab", "ba", "ac", "ca", "bc", "cb"]
             in do
                betterGuesser ["a", "b", "c"] DataMap.empty `shouldBe` Just (Guess "a")
                betterGuesser twoLetterWithoutRepeat (DataMap.fromList [(Guess "bc", Colors "10")]) `shouldBe` Just (Guess "ab")
                betterGuesser twoLetterWithoutRepeat (DataMap.fromList [(Guess "bc", Colors "11")]) `shouldBe` Just (Guess "cb")
                betterGuesser twoLetterWithoutRepeat (DataMap.fromList [(Guess "bc", Colors "01")]) `shouldBe` Just (Guess "ca")
                betterGuesser twoLetterWithoutRepeat (DataMap.fromList [(Guess "bc", Colors "02")]) `shouldBe` Just (Guess "ac")
                betterGuesser twoLetterWithoutRepeat (DataMap.fromList [(Guess "bc", Colors "20")]) `shouldBe` Just (Guess "ba")
                betterGuesser twoLetterWithoutRepeat (DataMap.fromList [(Guess "ab", Colors "00")]) `shouldBe` Nothing
                
            let threeLetterWords = [[x, y, z] | x <- ['a'..'c'], y <- ['a'..'c'], z <- ['a'..'c']] --- "aaa", "aab", ...
             in do
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aaa", Colors "000")]) `shouldBe` Just (Guess "bbb")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aaa", Colors "200")]) `shouldBe` Just (Guess "abb")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aaa", Colors "220")]) `shouldBe` Just (Guess "aab")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aab", Colors "000")]) `shouldBe` Just (Guess "ccc")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aab", Colors "100")]) `shouldBe` Just (Guess "cca")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aab", Colors "120")]) `shouldBe` Just (Guess "caa")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aab", Colors "001")]) `shouldBe` Just (Guess "bbc")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aab", Colors "101")]) `shouldBe` Just (Guess "bba")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aab", Colors "200")]) `shouldBe` Just (Guess "acc")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aab", Colors "201")]) `shouldBe` Just (Guess "abc")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aab", Colors "210")]) `shouldBe` Just (Guess "aca")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aab", Colors "211")]) `shouldBe` Just (Guess "aba")
                betterGuesser threeLetterWords (DataMap.fromList [(Guess "aaa", Colors "200"), (Guess "abb", Colors "220")]) `shouldBe` Just (Guess "abc")


src/Lib.hs
module Lib
    ( startGame
    , MasterType(..)
    , GuessType(..)
    , ColorsType(..)
    , judgeWord
    , simpleGuesser
    , betterGuesser
    , betterGuesserAll
    , checkMask
    , maskFunc
    , filterGuessByColor
    , countCharFreq
    , matchWord
    , checkFreqs
    , MaskType
    ) where

import qualified Data.Char as DataChar
import qualified Data.Maybe as DataMaybe
import qualified System.IO as SystemIO
import qualified Data.Map as DataMap
import qualified Data.List as DataList
import qualified System.Random as SystemRandom

newtype ColorsType  = Colors String deriving (Eq, Show)
newtype GuessType  = Guess String deriving (Show)
newtype MasterType = Master String
newtype ScoredGuessType = ScoredGuess (GuessType, Int)

type CandidateType = String
type CharFreq = DataMap.Map Char Int
type Guesser = History -> Maybe GuessType
type History = DataMap.Map GuessType ColorsType
type HistoryList = [(GuessType, ColorsType)]
type Judger  = GuessType -> ColorsType
type MaskType = String
type WordList = [String]

instance Eq GuessType where
  (Guess a) == (Guess b) = a == b

instance Ord GuessType where
  compare (Guess a) (Guess b) = compare a b

perfectMatch :: ColorsType
perfectMatch = Colors "22222"

startGame :: IO ()
startGame = do
  filteredWords <- filterWords . lines <$> readFile "./words.txt"
  zerosAndOnes <- SystemRandom.randomRs (0, 10000) <$> SystemRandom.newStdGen :: IO [Int]
  let history = DataMap.empty
  let shuffledWords = map snd $ DataList.sortBy (\(i, _) (c, _) -> compare (zerosAndOnes !! i) (zerosAndOnes !! c)) $ zip [0..] filteredWords

  putStrLn "Guide:"
  putStrLn "0     = black"
  putStrLn "1     = yellow"
  putStrLn "2     = green"
  putStrLn "Example:"
  putStrLn "22001 = 2 greens, 2 blacks, 1 yellow"
  putStrLn ""

  playTurn shuffledWords history
  putStrLn "Bye"

showHistoryItem :: (GuessType, ColorsType) -> String
showHistoryItem (Guess g, Colors c) = "History | guess: " ++ g ++ " " ++ c

playTurn :: WordList -> History -> IO ()
playTurn filteredWords history = do
  -- mapM (putStrLn) $ map showHistoryItem $ DataMap.toList history
  let newGuess = betterGuesser filteredWords history
  case newGuess of
    Nothing ->
      putStrLn "Nothing"
    Just (Guess g) ->
      putStr $ "Enter colors for '" ++ g ++ "': "
  SystemIO.hFlush SystemIO.stdout
  colors <- getLine
  case newGuess of
    Nothing ->
      putStrLn "Nothing found"
    Just g ->
      if Colors colors == perfectMatch then
        putStrLn "Found!"
      else
        let newHistory = DataMap.insert g (Colors colors) history
        in  playTurn filteredWords newHistory

filterWords :: WordList -> WordList
filterWords ws = map toLowerStr $ filter (\x -> 5 == length x) ws
  where
    toLowerStr :: String -> String
    toLowerStr = map DataChar.toLower

simpleScorer :: WordList -> History -> [ScoredGuessType]
simpleScorer words history =
  let isInHistory w = DataMap.member (Guess w) history
  in  map (f isInHistory) $ DataMap.toList history
    where
        f :: (String -> Bool) -> (GuessType, ColorsType) -> ScoredGuessType
        f isIncludedFunc (Guess g, Colors _)
            | isIncludedFunc g = ScoredGuess (Guess g, 0)
            | otherwise        = ScoredGuess (Guess g, 1)

simpleGuesser :: WordList -> History -> Maybe GuessType
simpleGuesser words history = fmap Guess (DataList.find (\w -> DataMap.notMember (Guess w) history) words)

betterGuesser :: WordList -> History -> Maybe GuessType
betterGuesser words history = DataMaybe.listToMaybe $ betterGuesserAll words history

betterGuesserAll :: WordList -> History -> [GuessType]
betterGuesserAll words history =
  map Guess $ filter (\w -> checkAgainstHistory w (DataMap.toList history)) words

checkAgainstHistory :: String -> HistoryList -> Bool
checkAgainstHistory candidate = foldl (\acc (g, c) -> acc && matchWord g c candidate) True

matchWord :: GuessType -> ColorsType -> CandidateType -> Bool
matchWord guess@(Guess gs) colors@(Colors cs) candidate =
  let eqs = maskFunc guess colors (== '2')
      nqs = maskFunc guess colors (/= '2')
      condGreens = checkMask eqs candidate (==)
      condOthers = checkMask nqs candidate (/=)
      cFreqs = countCharFreq (filterGuessByColor (/= '2') (Guess candidate) colors)
      yFreqs = countCharFreq (filterGuessByColor (=='1') guess colors)
      bFreqs = countCharFreq (filterGuessByColor (=='0') guess colors)
      condFreqs = checkFreqs cFreqs yFreqs bFreqs gs
  in
      condGreens && condOthers && condFreqs

checkFreqs :: CharFreq -> CharFreq -> CharFreq -> String -> Bool
checkFreqs cFreq yFreq bFreq = foldl f' True
  where
    f' acc g = let inCandidate = DataMaybe.fromMaybe 0 (DataMap.lookup g cFreq)
                   inYellows   = DataMaybe.fromMaybe 0 (DataMap.lookup g yFreq)
                   inBlacks    = DataMaybe.fromMaybe 0 (DataMap.lookup g bFreq)
                   ok = case (inYellows > 0, inBlacks > 0) of
                     (True,   True) -> inCandidate == inYellows
                     (False,  True) -> inCandidate == 0
                     (True,  False) -> inCandidate >= inYellows
                     (False, False) -> True
               in acc && ok

checkMask :: MaskType -> String -> (Char -> Char -> Bool) -> Bool
checkMask ms ws f = length ms == length ws && and (zipWith f' ms ws)
  where f' g w = g == '_' || f g w

filterGuessByColor :: (Char -> Bool) -> GuessType -> ColorsType -> String
filterGuessByColor f (Guess g) (Colors c) = map fst . filter (f . snd) $ zip g c

maskFunc :: GuessType -> ColorsType -> (Char -> Bool) -> MaskType
maskFunc (Guess g) (Colors c) f = zipWith (curry f') g c
  where f' (g, c) = if f c then g else '_'

judgeWord :: MasterType -> GuessType -> ColorsType
judgeWord (Master m) (Guess g) = Colors
                                . fst
                                . foldr toColorDigit ("", countCharFreq m)
                                $ zip m g
  where
    toColorDigit :: (Char, Char) -> (String, CharFreq) -> (String, CharFreq)
    toColorDigit (m, g) (acc, freqs) = (color:acc, decrCharFreq g freqs)
      where
        color :: Char
        color
              | DataMap.notMember g freqs = '0'
              | g == m = '2'
              | otherwise = '1'

decrCharFreq :: Char -> CharFreq -> CharFreq
decrCharFreq c = DataMap.filter (>0) . DataMap.insertWith (+) c (-1)

countCharFreq :: String -> CharFreq
countCharFreq word = DataMap.fromList
                    $ Prelude.map (\x->(head x, length x))
                    $ DataList.group
                    $ DataList.sort word
